feat(controller): add hostname exposing to units

Blystad opened the pull request:
This exposes a configurable hostname type to Docker container through
fleet unit files.

Three hostname types are supported:
- default     : standard Docker hostname
- application : application hostname (e.g. my-app.v54.web.1)
- server      : Core OS hostname

This fixes #2835 and is an alternative to #2428, which got rejected.

It is necessary when using a monitoring solution like New Relic.

deis-admin commented:
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

carmstrong commented:
Jenkins, add to whitelist.

carmstrong commented:
I think this is a great compromise for hostnames! My only comment would be that we typically don't include notes like these in the Customizing Deis section, but I honestly can't think of a better location for the information in this PR.

I would also add an example for setting this with ``deisctl``, but otherwise this looks great!

carmstrong commented:
@Blystad Jenkins is failing with a style error in `fleet.py`. I confirmed that this isn't a problem on master:

```console
➜  controller git:(pr-2837) venv/bin/flake8
./scheduler/fleet.py:101:1: C901 'FleetHTTPClient._create_container' is too complex (14)
➜  controller git:(pr-2837) git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
➜  controller git:(master) git pull
Already up-to-date.
➜  controller git:(master) venv/bin/flake8
➜  controller git:(master)
```

carmstrong commented on controller/scheduler/fleet.py:
```diff
@@ -112,6 +113,18 @@ def _create_container(self, name, image, command, unit, **kwargs):
+            l.update({'hostname': ''})
+        elif hostname == "application":
+            # replace underscore with dots, since underscore is not valid in DNS hostnames
+            dns_name = name.replace("_", ".")
```
Maybe we should replace underscores with hyphens instead? I'm thinking dots typically imply subdomains, which may cause issues with some applications. Or maybe I'm entirely off-base.

Blystad replied:
I considered hyphens, and I considered stripping it out entierly. I can't see how that could cause issues with some applications, and also it would make filtering easier as it will be easier to write filters (when searching through logs which includes the hostname) where `my-application.*.web.*` is found.

Blystad commented:
Added `deisctl` example and fixed the complexity by moving the hostname functionality into it's own function, which is better anyways. Now I'll just wait for Jenkins to pass.

Blystad commented:
It passed, and now I see a "Confirm Merge" button. Should I click that?

carmstrong commented:
Code LGTM! We're shipping Deis v1.2.0 on Tuesday, so this will likely merge shortly thereafter.

bacongobbler commented:
this is really well designed. Nice job @Blystad!

bacongobbler commented on docs/customizing_deis/controller_settings.rst:
```diff
@@ -73,3 +74,38 @@ Be sure that your custom image functions in the same way as the `stock controlle
+
+The valid ``unitHostname`` alternatives are:
+
+default
```
Could we please make this a [definition list](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#definition-lists) instead?

Blystad commented:
@bacongobbler Thanks for the kind words. I've changed the documentation into a definition list now.

bacongobbler commented:
code LGTM. :shipit:
